### PR TITLE
EarlGrey 2.0 Prototype (multiprocess, XCUITest support)

### LINKED ISSUE #63: Support XC UI Testing
Integrate EarlGrey with XC UI Testing. Please follow this issue for progress updates.

### PR DESCRIPTION
**Pull request by axi0mX**
Here's something I made this summer. You can look over the tests to understand how to use it, and read the code to understand how it works.

I didn't add any tests using XCUITest API, feel free to add your own.

I also wrote a doc about it, feel free to comment on it:
https://docs.google.com/document/d/1nlJKiPvW4s_4XkeV9fMP-s90FoPAuGRTBoHCyz3F2nE/edit?usp=sharing

It's just a prototype, so expect to find bugs.

Resolves #63.

### CODE REVIEW

**Comment by bootstraponline** [comment]
Do we need IPC if we can just link in [Facebook's XCUITest library?](https://github.com/google/EarlGrey/issues/55#issuecomment-226219297)

**Comment by axi0mX** [comment]
If your only goal is to use XCUITest and EarlGrey together, no. 

Multiprocess design has other advantages though, not just XCUITest support. With it, the prototype can today inject EarlGrey into SpringBoard on iOS simulator and confirm system alert dialogs using EarlGrey API. 

In the future it can be extended to inject EarlGrey into any application on iOS simulator and on jailbroken iOS devices, and allow testing of Today widgets, Notifications, iMessage extensions, Siri extensions, Maps extensions, other app extensions, and anything else that requires testing in apps that you cannot compile and deploy yourself such as SpringBoard, Calendar, Photos, and Mobile Safari. All of this within a single test.

File: EarlGrey/Interprocess/GREYApplication.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+// Objective-C only. Disabled in Swift because block syntax doesn't work with Objective-C blocks.
+- (void)executeSyncWithBlock:(GREYTestBlock)block NS_SWIFT_UNAVAILABLE("Use executeSync: instead");
+// Swift only. Not for use in Objective-C.
+- (void)executeSync:(GREYTestFunction)function;
+
+- (void)launch;
+- (void)terminate;
+
+- (BOOL)isHealthy;
```
Comment (tirodkar) [inline EarlGrey/Interprocess/GREYApplication.h]: It would be nice to have this documented.

**Comment by Liusunming** [comment]
Is real device supported to confirm system alert dialogs using EarlGrey API?

**Comment by Liusunming** [comment]
Is real device supported to confirm system alert dialogs using EarlGrey API?

**Comment by tirodkar** [comment]
@Liusunming the current prototype being worked out will be using XCUITest API's to work with System Dialogs on sim/device.

**Comment by Liusunming** [comment]
@tirodkar, does it only work within UI testing target, but not Unit testing target? In the testSystemDialog example, it just return if not on simulator:

  // Interaction with SpringBoard is not on device or before iOS 9.
  if (!TARGET_OS_SIMULATOR || !iOS9_OR_ABOVE()) {
    return;
  }

**Comment by tirodkar** [comment]
@Liusunming Yes, EarlGrey 2.0 will only work in the XCUITestCase context.

**Comment by Liusunming** [comment]
@tirodkar, thank you so much.
 In #issue55, bootstraponline worked WebDriverAgentLib with EarlGrey in unit testing target. I don't know how it works.

I just got the Facebook lib working with EarlGrey and I think it's quite powerful.

Built WebDriverAgentLib.framework for release
Added to Link Binary with Libraries
Updated bridging header:
#import <EarlGrey/EarlGrey.h>

#import <WebDriverAgentLib/WebDriverAgentLib.h>
#import <WebDriverAgentLib/XCUIApplication.h>
Now I have full access to all the XCUITest APIs, including the Facebook versions (which are almost always better). Including the ability to deal with alerts.
let alert = FBAlert(application: XCUIApplication.init(privateWithPath: nil, 
                                                             bundleID: "my.app.bundle"))
alert.isPresent()
This is all within the unit testing target, no need for a multi-process model.

**Comment by Liusunming** [comment]
@tirodkar thank you so much. I used EarlGrey together with WebDriverAgentLib to confirm system alert dialogs.

**Comment by wuhao5** [comment]
@Liusunming do you mind sharing more info about how you do it?

**Comment by Liusunming** [comment]
Use FBAlert of WebDriverAgentLib with EarlGrey2.0 in XCUITestCase.

**Comment by eder7** [comment]
Hi, is there a timeline for Earl Grey 2.0? Will you be able to use XCUITest calls like `app.buttons["2"].tap()`?

**Comment by tirodkar** [comment]
@eder7 We're currently developing 2.0. As soon as we have it vetted internally, we'll have it announced here.

We're keeping the current EarlGrey syntax, so instead of `app.buttons["2"].tap()`, you would have to do `EarlGrey.select(elementWithMatcher:grey_buttonTitle("w")).perform(grey_tap())`. The current API will pretty much remain the same for 2.0 as well.

Is there any other particular XCUITest api that you need to use?

**Comment by khandpur** [comment]
To answer @eder7 question: Yes you would be able to because after all it's running within an XCUITest.

**Comment by Liusunming** [comment]
@khandpur but it is not using EarlGrey's property if you are using app.buttons["2"].tap().

**Comment by khandpur** [comment]
That's right. The test is run in the same context so user can choose to use xcuitest api's directly but will lose all the features offered by EarlGrey.

**Comment by khandpur** [comment]
EarlGrey 2 does this!
https://github.com/google/EarlGrey/tree/earlgrey2

**Pull request closed without merging**